### ISSUE TITLE
selector-pseudo-colon-no-invalid-space

### ISSUE THREAD

**Issue by evilebottnawi** [stylelint/stylelint#1255]
> Describe the issue. Is it a bug or a feature request (new rule, new option, etc.)?

Valid:
```css
a:hover {
  color: red;
}
```

```css
a
:hover {
  color: red !important;
}
```

With one options: "never"

**Comment by evilebottnawi** [comment]
Spaces before always are invalid

**Comment by davidtheclark** [comment]
How about `selector-pseudo-colon-no-space-around`.

If there's just a "never" option, seems to me it should be a `-no-` rule.

**Comment by evilebottnawi** [comment]
@davidtheclark sound good
@jeddy3 What do you think?

**Comment by davidtheclark** [comment]
If you or anyone wants to start on this feel free. @jeddy3 can comment on the name whenever he's back.

**Comment by evilebottnawi** [comment]
@davidtheclark i think we should add options `"ignore": ["nested"]` for
```scss
a {
  :hover,
  :focus {
    color: red;
  }
}
```

**Comment by davidtheclark** [comment]
Yeah, makes sense.

**Comment by evilebottnawi** [comment]
@jeddy3 @davidtheclark we should revise the rule, because (with space after)
```css
a ::after {
    content: "1";
    color: red;
}
```
is valid construction for pseudo-elements.

**Comment by evilebottnawi** [comment]
With pseudo classes always invalid space before and after. May be rename to selector-pseudo-classes-colon-no-space-around?

**Comment by davidtheclark** [comment]
Or we could just make `selector-pseudo-colon-no-space-after`

**Comment by evilebottnawi** [comment]
@davidtheclark for pseudo-classes invalid and before. Maybe
1. `selector-pseudo-classes-colon-no-space-before`
2. `selector-pseudo-classes-colon-no-space-after`
3. `selector-pseudo-element-colon-no-space-after`
?

**Comment by davidtheclark** [comment]
Might be simpler with:
- `selector-pseudo-colon-no-space-after`
- `selector-pseudo-classes-colon-no-space-around`

**Comment by evilebottnawi** [comment]
@davidtheclark `selector-pseudo-classes-colon-no-space-around` sound good, but `selector-pseudo-colon-no-space-after` may duplicate errors from the previous rules.

**Comment by jeddy3** [comment]
```css
a ::after {
    content: "1";
    color: red;
}
```

Is that really valid? Wow.

Sounds like there's three things going on here.

1. Disallowing *invalid* spaces in *pseudo-classes*.
2. Disallowing *invalid* spaces in *pseudo-elements*.
2. Allowing a user to specify a space or no space before the colon in *pseudo-elements*.

Does that sound about right? If so, does the following sound?:

- `selector-pseudo-class-colon-no-invalid-space: true` (checks for space before and after)
- `selector-pseudo-element-colon-no-invalid-space: true` (checks for space after)
- `selector-pseudo-element-colon-space-before: "always"|"never"

>  i think we should add options "ignore": ["nested"]

Is this even valid?:

```scss
a {
  :hover,
  :focus {
    color: red;
  }
}
```

I thought you had to use the *nesting selector* (even in SCSS) in this situation e.g.

```scss
a {
  &:hover,
  &:focus {
    color: red;
  }
}
```

**Comment by evilebottnawi** [comment]
@jeddy3 valid in SCSS, we get
```css
a :hover,
a :focus {
  color: red;
}
```
And non valid in CSS.

**Comment by davidtheclark** [comment]
@evilebottnawi I don't think that's valid. I tried it and it compiles to the same thing: `a :hover {}` -> `a :hover {}` — which means it's outputting invalid CSS.

@jeddy3 I think the space before the pseudo-element is "valid" because it just means "Target the `::before` pseudo-elements", not that it's the equivalent of having no space. `a ::before {}` !== `a::before {}`. Right, @evilebottnawi?

**Comment by evilebottnawi** [comment]
@davidtheclark right

**Comment by evilebottnawi** [comment]
@davidtheclark but, if who used this construction, usually made mistake, where rarely used

**Comment by davidtheclark** [comment]
I like `selector-pseudo-colon-no-space-after` because that's always invalid.

How about for the other situation a rule like `selector-no-simple-pseudo`, or a better name? This would disallow either pseudo type with a space before it.

**Comment by evilebottnawi** [comment]
@davidtheclark 
Then may be we add option `ignorePseudoElement` (or better name), for disable checks space before pseudo elements?

**Comment by jeddy3** [comment]
How's about?:

- `selector-pseudo-colon-no-invalid-space` - as we use "invalid" when checking for invalid code. Without the word "invalid" we're normally specifying a preference between two "valid" alternatives e.g. `selector-no-type`.
- `selector-pseudo-no-unqualified` - as we've used "qualified" to mean "prefixed" with a simple selector else where.

The latter rule is weird to me though as they're odd exceptions e.g. `::selection`. Can we make this issue about the *invalid* space rule and tackle the other one another time?

**Comment by davidtheclark** [comment]
Sure those rule names SGTM.

**Comment by jeddy3** [comment]
- Name: `selector-pseudo-colon-no-invalid-space`
- Primary options: `true`
- Secondary options: none
- Message: `Unexpected invalid space `
- Section: "Selector"

**Comment by evilebottnawi** [comment]
@jeddy3 can you give example code for this rules with comments? I'm a little confused.

**Comment by jeddy3** [comment]
```css
a: hover {} /* warning as invalid space after pseudo-class  */
a :hover {} /* warning as invalid space before pseudo-class */
a:: before {} /* warning as invalid space after pseudo-element */
a ::before {} /* NO warning if space after pseudo-element as could be implied universal selector e.g. a *::above {} */
```

**Comment by evilebottnawi** [comment]
@jeddy3 thanks, how about add rule allow disable `a ::before` lonely pseudo-elements?

**Comment by jeddy3** [comment]
Sure, but lets talk about that in a separate issue.

**Comment by jeddy3** [comment]
After looking through the specs more closely, I've a few misgivings about this rule now. 

1. "pseudo" isn't really a *thing*. There are *pseudo-classes* and *pseudo-elements*, but not "pseudos". We've only targeted the latter two elsewhere.
2. I'm not entirely convinced we're right about where space is valid or invalid.  There's *a lot* going on in [this spec](https://www.w3.org/TR/selectors4/#the-lang-pseudo).

As such, I'm in favour of pushing this group of rules to experimental plugins.

**Comment by evilebottnawi** [comment]
@jeddy3 tomorrow i think about this rules and i came to this conclusion - be good use rules `selector-pseudo-class-no-lonely` and `selector-pseudo-element-no-lonely`(with ignore option). Why?
Get above example:
```css
a: hover {} /* We have `Error: Unexpected "space" found.` and it is right, even if not get error rule `selector-type-no-unknown` throw error about unknown `hover` */
a :hover {} /* Good for rule `selector-pseudo-class-no-lonely` */
a:: before {} /* Also we have `Error: Unexpected "space" found.` and it is right, even if not get error rule `selector-type-no-unknown` throw error about unknown `before` */
a ::before {} /* Good for rule `selector-pseudo-element-no-lonely` */
a ::strange-pseudo-element /* Good for rule `selector-pseudo-element-no-lonely` with option `ignore: ['strange-pseudo-element']` */
::selection /* Default ignore by `selector-pseudo-element-no-lonely` */
::unknown /* `selector-pseudo-element-no-unknown` throw error */
```css

**Comment by jeddy3** [comment]
Lets push this to an experimental plugin for now as it raises too many questions.

Feel free build and provide a home for this plugin yourself, use it on your code base, and add it to the plugins list under an "Experimental" section.

**Comment by evilebottnawi** [comment]
@jeddy3 What do you think about name this rule?

**Comment by jeddy3** [comment]
I'd probably name the plugins `stylelint-selector-pseudo-element-no-unqualified` and `stylelint-selector-pseudo-class-no-unqualified` - as we've used "qualified" to mean "prefixed" with a simple selector else where.
